### PR TITLE
Updates to bounding box interface and cleanup

### PR DESCRIPTION
**Pull request by embray**
This is the update to #4040--it includes all the fixes from #4040, but also includes my updates to the bounding box interface.

The purpose of these updates, as explained in the commit message of 779ef3e5bb14d9711dc24ae26578bfd7c22d618b, is to make the interface for `bounding_box` fairly consistent with the interface for `Model.inverse`.  To that end the `bounding_box_default` method is eliminated, and model classes that define a bounding box can do so as a class-level attribute, method, or property simply called `bounding_box`, which is then wrapped by the Model metaclass to incorporate the default getter/setter behavior for the `Model.bounding_box` property.  This is similar to how `Model.inverse` works.

One downside to this that I just noticed while testing the documentation is that if a `Model` subclass has a `bounding_box` with its own docstring, that docstring usually gets thrown away.  We definitely don't want that, so I am going to open a separate issue to fix that later.

### CODE REVIEW

**Comment by embray** [comment]
cc: @patti @adonath

**Comment by patti** [comment]
:+1:

File: astropy/modeling/core.py [hunk 1]
```diff
@@ -940,6 +1073,106 @@ def bounding_box(self, limits):
...
+                out = np.zeros(coords[0].shape)
+
+        if out is not None:
+            try:
+                assert out.ndim == ndim
+            except AssertionError:
+                raise AssertionError(
+                    'The array and model must have the same number '
+                    'of dimensions.')
+
```
Comment (embray) [inline astropy/modeling/core.py]: At some point something needs to be done with these assertions--really there's no reason to use assert and then catch AssertionErrors and reraise them.  This should just raise a ValueError something after checking the condtion, and not use `assert` at all.  I don't think I have time to fix it now though.

**Comment by embray** [comment]
FWIW the Sphinx build for this failed due to a failure to fetch the matplotlib object.inv.  Didn't appear to have any failures actually related to PR, but I've restarted it anyways.

File: docs/modeling/bounding-boxes.rst [hunk 2]
```diff
@@ -13,137 +14,145 @@ image from a large survey catalog, each source should only be evaluated over the
...
-------------------------------------------
+.. warning::
 
-When a model is evaluated over a range much larger than the model itself, it may
-be prudent to use `astropy.modeling.render_model` if efficiency is a concern.
-The ``render_model`` function can be used to evaluate a model on an array of the
-same dimensions. If no array is given, ``render_model`` will return a "postage
-stamp" array corresponding to the bounding box region. However, if
-``bounding_box`` is `None` an image or coordinates must be passed.
+    Note: Currently when creating a new compound model by combining multiple
```
Comment (eteq) [inline docs/modeling/bounding-boxes.rst]: "Note:" is not necessary in a warning (and perhaps confusing, because ``.. note::`` means something different)

Comment (astrofrog) [inline docs/modeling/bounding-boxes.rst]: Fixed in 96f531b3f1995e5d2f30a78d2a0261499d41477a

Comment (embray) [inline docs/modeling/bounding-boxes.rst]: :+1:

**Pull request merged**